core/workflow-operator/src/main/scala/edu/uci/ics/amber/operator/sink/managed/ProgressiveSinkOpDesc.java

```diff
@@ -95,8 +101,16 @@ public PhysicalOp getPhysicalOp(WorkflowIdentity workflowId, ExecutionIdentity e
 
                             javaMap.put(operatorInfo().outputPorts().head().id(), outputSchema);
 
-                            // set schema for the storage
-                            getStorage().setSchema(outputSchema);
+                            // the storage metadata is reset if it is a MongoDocument, because the output schema is calculated here
+                            // TODO: after we change the execution as:
+                            //   1. propagate all the schemas
+                            //   2. assign op result storage
+                            //   we can clean up this logic
+                            if (this.storage instanceof MongoDocument) {
+                                MongoDocument<Tuple> storage = (MongoDocument<Tuple>)getStorage();
+                                storage.setDeserde(Tuple.fromDocument().apply(outputSchema));
+                                this.opResultStorage.setSchema(operatorIdentifier(), outputSchema);
+                            }
```

Comment:
this logic should be moved after schema propagation. not to be inside propagation function.